File: app/contact_management/client_autocomplete.js
// Query text stays in the browser. Only an explicit result choice sends an event.
export function rankedContacts(data, query) {
  const q = Array.from(query.trim()).map(c => data.folds[c] ?? c.toLowerCase()).join("");
  if (!q) return [];
  return data.contacts.map((contact, order) => {
    const [first, last, full, company, email] = contact.fields;
    const matches = [first.startsWith(q), last.startsWith(q), full.startsWith(q),
      first.includes(q), last.includes(q), full.includes(q),
      company.startsWith(q), company.includes(q), email.startsWith(q), email.includes(q)];
    return {contact, order, priority: matches.indexOf(true)};
  }).filter(item => item.priority >= 0)
    .sort((a, b) => a.priority - b.priority || a.order - b.order).map(item => item.contact);
}

export default function({parentElement, data, setTriggerValue}) {
  const input = parentElement.querySelector("input");
  const menu = parentElement.querySelector('[role="listbox"]');
  if (input.dataset.selectedId !== data.selected_id) {
    input.value = "";
    input.dataset.selectedId = data.selected_id;
  }
  let results = [];
  let active = -1;
  function close() {
    menu.hidden = true;
    input.setAttribute("aria-expanded", "false");
    input.removeAttribute("aria-activedescendant");
  }
  function choose(contact) {
    input.value = "";
    close();
    setTriggerValue("selected", contact.id);
  }
  function show() {
    results = rankedContacts(data, input.value);
    active = -1;
    menu.replaceChildren();
    input.removeAttribute("aria-activedescendant");
    if (!input.value.trim()) return close();
    menu.hidden = false;
    input.setAttribute("aria-expanded", "true");
    if (!results.length) {
      const empty = document.createElement("div");
      empty.className = "empty";
      empty.setAttribute("role", "status");
      empty.textContent = "No matching client contacts.";
      menu.append(empty);
    }
    results.forEach((contact, index) => {
      const option = document.createElement("div");
      option.id = `contact-option-${index}`;
      option.setAttribute("role", "option");
      option.setAttribute("aria-selected", "false");
      option.textContent = contact.label;
      option.onmousedown = event => event.preventDefault();
      option.onclick = () => choose(contact);
      menu.append(option);
    });
  }
  input.oninput = show;
  input.onfocus = show;
  input.onblur = close;
  input.onkeydown = event => {
    if (event.key === "Escape") return close();
    if (event.key === "ArrowDown" || event.key === "ArrowUp") {
      event.preventDefault();
      if (menu.hidden) show();
      if (!results.length) return;
      active = (active + (event.key === "ArrowDown" ? 1 : -1) + results.length) % results.length;
      Array.from(menu.children).forEach((option, index) => option.setAttribute("aria-selected", String(index === active)));
      input.setAttribute("aria-activedescendant", menu.children[active].id);
      menu.children[active].scrollIntoView({block: "nearest"});
    } else if (event.key === "Enter" && !menu.hidden && active >= 0) {
      event.preventDefault();
      choose(results[active]);
    }
  };
  close();
}

File: app/contact_management/client_autocomplete.py
"""Browser-local contact finder using the installed Streamlit component API."""
from pathlib import Path

import streamlit as st

from app.contact_management.models import ClientContact


_component = st.components.v2.component(
    "photography_client_autocomplete",
    html='''<label for="contact-search">Client Contact Search</label>
    <input id="contact-search" role="combobox" aria-autocomplete="list"
      aria-controls="contact-results" aria-expanded="false" autocomplete="off"
      placeholder="Search by company, name, or email...">
    <div id="contact-results" role="listbox" hidden></div>''',
    css='''label { display:block; margin-bottom:.4rem; font-size:.875rem; }
    input { box-sizing:border-box; width:100%; padding:.65rem; border:1px solid #8886;
      border-radius:.5rem; font:inherit; color:var(--st-text-color);
      background:var(--st-secondary-background-color); }
    #contact-results { max-height:240px; overflow-y:auto; border:1px solid #8886;
      border-radius:.4rem; background:var(--st-background-color); }
    [role=option] { padding:.6rem; cursor:pointer; overflow-wrap:anywhere; }
    [role=option]:hover, [aria-selected=true] { background:var(--st-secondary-background-color); }
    .empty { padding:.6rem; }''',
    js=Path(__file__).with_suffix(".js").read_text(encoding="utf-8"),
)


def autocomplete_data(contacts: list[ClientContact], selected_id: str) -> dict:
    ordered = sorted((c for c in contacts if c.active), key=lambda c: (
        c.full_name.casefold(), c.company_name.casefold(), c.email.casefold(), c.id))
    fields = [(c.first_name, c.last_name, c.full_name, c.company_name, c.email) for c in ordered]
    characters = set("".join(value for row in fields for value in row))
    characters.update("".join(char.upper() + char.lower() for char in list(characters)))
    return {
        "selected_id": selected_id,
        "folds": {char: char.casefold() for char in characters},
        "contacts": [{"id": c.id, "label": c.dropdown_label,
                      "fields": [value.casefold() for value in row]}
                     for c, row in zip(ordered, fields)],
    }


def apply_autocomplete_selection(value: object, contacts: list[ClientContact]) -> None:
    if isinstance(value, str) and any(c.active and c.id == value for c in contacts):
        st.session_state["photo_pricing_client_contact_id"] = value


def render_client_autocomplete(contacts: list[ClientContact], selected_id: str) -> None:
    revision = st.session_state.get("photo_pricing_client_search_revision", 0)
    key = f"photo_pricing_client_autocomplete_{revision}"

    def select_contact() -> None:
        apply_autocomplete_selection(st.session_state[key].get("selected"), contacts)

    _component(key=key, data=autocomplete_data(contacts, selected_id), on_selected_change=select_contact)
